modules/server/src/main/scala/scaladex/server/service/UserSessionSynchronizer.scala

```diff
@@ -0,0 +1,58 @@
+package scaladex.server.service
+
+import java.util.UUID
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+
+import akka.actor.ActorSystem
+import cats.implicits.toTraverseOps
+import scaladex.core.model.UserState
+import scaladex.core.service.GithubService
+import scaladex.core.service.SchedulerDatabase
+import scaladex.infra.GithubClient
+
+class UserSessionSynchronizer(database: SchedulerDatabase)(implicit ec: ExecutionContext, ac: ActorSystem)
+    extends Scheduler("user-session-synchronizer", 1.hour) {
+
+  type Session = (UUID, UserState)
+
+  override def run(): Future[Unit] =
+    for {
+      sessions <- database.getAllSessions()
+      sessionsWithClients = createClientsForEach(sessions)
+      updatedEitherUserSessions <- sessionsWithClients.traverse {
+        case (session, client) => updateUserSession(session, client)
+      }
+      expiredSessionIds = updatedEitherUserSessions.collect { case Left(expiredSessionId) => expiredSessionId }
+      updatedUserSessions = updatedEitherUserSessions.collect { case Right(sessions) => sessions }
+      _ <- updatedUserSessions.traverse { case (userId, userState) => database.insertSession(userId, userState) }
+      _ <- expiredSessionIds.traverse(database.deleteSession)
+    } yield ()
+
+  private def createClientsForEach(sessions: Seq[Session]): Seq[(Session, GithubService)] =
+    sessions.map { case session @ (_, userState) => (session, new GithubClient(userState.info.token)) }
+
+  private def updateUserSession(session: Session, service: GithubService): Future[Either[UUID, Session]] =
+    service
+      .getUserInfo()
+      .map { updatedUserInfo =>
+        session match {
+          case (userId, staleUserState) => Right((userId, staleUserState.copy(info = updatedUserInfo)))
+        }
+      }
+      .recoverWith {
+        case _ =>
```

Comment:
> I think GithubService#getUserInfo doesn't fail in a way that lets the consumer know why something failed, but perhaps I'm misunderstanding something.

That's right.

We cannot assume that all failures are due to revoked tokens because we can also have some network outage or the Github API could be unstable. In those cases we don't want the scheduler to delete all the sessions.

So we need to change the implementation of `GithubService#getUserInfo`.
I see two possible solutions:
1. In https://github.com/scalacenter/scaladex/blob/91932e3d98b3262f2ad1d1201a95e2e30fb76e10/modules/infra/src/main/scala/scaladex/infra/GithubClient.scala#L368 we throw a custom `HttpException(code, reason)` that we can then catch and inspect.
2. Or, we change the signature and implementation of  `GithubService#getUserInfo` to return a `Future[GithubResponse[UserInfo]]`

I let you choose the one you prefer.